lib/ansible/modules/packaging/os/snap.py

```diff
@@ -0,0 +1,189 @@
+
+from __future__ import absolute_import, division, print_function
+__metaclass__ = type
+
+ANSIBLE_METADATA = {
+    'metadata_version': '1.1',
+    'status': ['preview'],
+    'supported_by': 'community'
+}
+
+DOCUMENTATION = '''
+---
+module: snap
+
+short_description: Manages snaps
+
+version_added: "2.8"
+
+description:
+    - "Manages snaps packages."
+
+options:
+    name:
+        description:
+            - Name of the snap to install or remove.
+        required: true
+    state:
+        description:
+            - Desired state of the package.
+        required: false
+        default: present
+        choices: [ absent, present ]
+    classic:
+        description:
+            - Confinement policy. The classic confinment allows a snap to have
+              the same level of access to the system as "classic" packages,
+              like those managed by APT. This option corresponds to the --classic argument.
+        type: bool
+        required: false
+        default: False
+
+author:
+    - Victor Carceler (vcarceler@iespuigcastellar.xeill.net)
+    - Stanislas Lange (angristan) <angristan@pm.me>
+'''
+
+EXAMPLES = '''
+# Install "foo" snap
+- name: Install foo
+  snap:
+    name: foo
+
+# Remove "foo" snap
+- name: Remove VLC
+  snap:
+    name: foo
+    state: absent
+
+# Install a snap with classic confinement
+- name: Install "foo" with option --classic
+  snap:
+    name: foo
+    classic: True
+'''
+
+RETURN = '''
+msg:
+    description: An informative message on the task's output
+    type: str
+    returned: on error
+'''
+
+from ansible.module_utils.basic import AnsibleModule
+
+
+def clean_err(err):
+    # Remove whitespaces and newlines that the snap CLI outputs.
+    return err.replace('\n', '').replace('      ', '').replace('"', '\'')
+
+
+def snap_exists(module, snap_name):
+    snap_path = module.get_bin_path("snap", True)
+    cmd = "%s info %s" % (snap_path, snap_name)
+    rc, out, err = module.run_command(cmd, check_rc=False)
+
+    return rc, out, err
+
+
+def is_snap_installed(module, snap_name):
+    snap_path = module.get_bin_path("snap", True)
+    cmd = "%s list %s" % (snap_path, snap_name)
+    rc, out, err = module.run_command(cmd, check_rc=False)
+
+    return rc, out, err
+
+
+def install_snap(module, snap_name):
+    rc, out, err = is_snap_installed(module, snap_name)
+    if rc == 0:
+        # Snap is already installed
+        return False
+
+    if module.check_mode:
+        module.exit_json(changed=True)
+
+    classic = '--classic' if module.params['classic'] else ''
+
+    snap_path = module.get_bin_path("snap", True)
+    cmd = "%s install %s %s" % (snap_path, snap_name, classic)
+
+    # Actually install the snap
+    rc, out, err = module.run_command(cmd, check_rc=False)
+
+    if rc == 0:
+        # Snap has been installed
+        return True
+    else:
+        # Something went wrong
+        err = clean_err(err)
```

Comment:
It is more useful to extract some status data from the message and prepend some easy-to-understand explanation, this abstracts user from needing to actually know the detail of underlying command implementation.
Also, they should be able to access stderr/out/_lines via the result variable if needed anyway.

Reply:
In the case of an error message like this:

```sh
root@server:~# snap install micro
error: This revision of snap "micro" was published using classic confinement and thus may perform
       arbitrary system changes outside of the security sandbox that snaps are usually confined to,
       which may put your system at risk.

       If you understand and want to proceed repeat the command including --classic.
```

Which currently outputs like this:

```yml
fatal: [host]: FAILED! => {"changed": false, "msg": "error: This revision of snap 'micro' was published using classic confinement and thus may perform arbitrary system changes outside of the security sandbox that snaps are usually confined to, which may put your system at risk. If you understand and want to proceed repeat the command including --classic."}
```

What can I do ?cThere is no way for me to know in advance that the snap requires the classic confinement.

What I can do is to set `msg = "The snap xxx couldn't be installed"` and `err = clean_err(err)` (and `out = clean_err(err)`?)

Reply:
Also, stdout when snap installs a pakages is really messy (you should try yourself, it's like when you do a `npm install`, there is only one line that is updated, not lines appending to each other like `apt-get`)

It can look like this:

```yml
ok: [host] => {"changed": false, "msg": "Name     Version                  Rev  Tracking  Publisher  Notes\nspotify  1.0.88.353.g15c26ea1-14  19   stable    spotify*   -\n", "stderr": "", "stderr_lines": [], "stdout": "Name     Version                  Rev  Tracking  Publisher  Notes\nspotify  1.0.88.353.g15c26ea1-14  19   stable    spotify*   -\n", "stdout_lines": ["Name     Version                  Rev  Tracking  Publisher  Notes", "spotify  1.0.88.353.g15c26ea1-14  19   stable    spotify*   -"]}
```

I can't even clean that up lol. Should I do like before, setting `msg` to `"Snaps xxx, xxx, xxx have been installed"`?

Reply:
See https://github.com/ansible/ansible/pull/44939/commits/44b7ca0651fff18f0fb1cb880d06509666780cc5

Reply:
I've added some comments to PR.
If you have some docker image in mind where this snap is pre-installed, I can try.
Regarding sanitization of term output, it might be messy, yes. But let's look at this successful output:
```python
cmd_stdout = """Name     Version                  Rev  Tracking  Publisher  Notes
spotify  1.0.88.353.g15c26ea1-14  19   stable    spotify*   -
"""

# cut the headers line and drop the empty trailer:
packages_output = cmd_stdout.partition('\n')[-1].strip()

# split lines and clean up trailing and starting spaces:
packages_outputs = list(map(str.strip, packages_output.split('\n')))

for package_line in packages_outputs:
    # of course, the regexp deserves to be adjusted to match format
    m = re.match(r'(?P<package_name>\w+)\s+(?P<package_version>\w+)\s+(?P<package_rev>\w+)\s+(?P<package_tracking>\w+)\s+(?P<package_publisher>\w+)\s+(?P<notes>\w+)\s+', package_line)
    if m is None:
        continue
    # now, we have access to all fields of the snap:
    m.group('package_name')
    m.group('package_version')
    m.group('package_rev')
    m.group('package_tracking')
    m.group('package_publisher')
    m.group('notes')
```

Reply:
Snap requires a daemon, thus an init system, so you won't be able to test it in Docker.

Reply:
We use docker for our CI, and run init systems already, so I don't see a reason why we wouldn't be able to do so here too.  We have ubuntu images where this should be achievable.

Reply:
What I mean is that snap is composed of a daemon and CLI tool. The daemon is a service so it depends on sysvinit/systemd.

```sh
stanislas@mbp ~> docker run -it --name snap ubuntu:18.04
root@d908a2194478:/# snap
bash: snap: command not found
root@d908a2194478:/# apt-get update -qq && apt-get install snapd
Reading package lists... Done
Building dependency tree
Reading state information... Done
The following additional packages will be installed:
  apparmor ca-certificates dbus dirmngr dmsetup file gir1.2-glib-2.0 gnupg gnupg-l10n gnupg-utils gpg gpg-agent gpg-wks-client gpg-wks-server gpgconf gpgsm krb5-locales
  libapparmor1 libargon2-0 libasn1-8-heimdal libassuan0 libbsd0 libcap2 libcryptsetup12 libdbus-1-3 libdevmapper1.02.1 libedit2 libexpat1 libfuse2 libgirepository-1.0-1
  libglib2.0-0 libglib2.0-data libgssapi-krb5-2 libgssapi3-heimdal libhcrypto4-heimdal libheimbase1-heimdal libheimntlm0-heimdal libhx509-5-heimdal libicu60 libidn11
  libip4tc0 libjson-c3 libk5crypto3 libkeyutils1 libkmod2 libkrb5-26-heimdal libkrb5-3 libkrb5support0 libksba8 libldap-2.4-2 libldap-common liblzo2-2 libmagic-mgc
  libmagic1 libmpdec2 libnpth0 libnss-systemd libpam-systemd libpython3-stdlib libpython3.6-minimal libpython3.6-stdlib libreadline7 libroken18-heimdal libsasl2-2
  libsasl2-modules libsasl2-modules-db libsqlite3-0 libssl1.0.0 libssl1.1 libwind0-heimdal libx11-6 libx11-data libxau6 libxcb1 libxdmcp6 libxext6 libxml2 libxmuu1
  mime-support multiarch-support networkd-dispatcher openssh-client openssl pinentry-curses python3 python3-dbus python3-gi python3-minimal python3.6 python3.6-minimal
  readline-common shared-mime-info squashfs-tools systemd systemd-sysv udev xauth xdg-user-dirs xz-utils
Suggested packages:
  apparmor-profiles-extra apparmor-utils default-dbus-session-bus | dbus-session-bus dbus-user-session pinentry-gnome3 tor parcimonie xloadimage scdaemon fuse krb5-doc
  krb5-user libsasl2-modules-gssapi-mit | libsasl2-modules-gssapi-heimdal libsasl2-modules-ldap libsasl2-modules-otp libsasl2-modules-sql iw | wireless-tools keychain
  libpam-ssh monkeysphere ssh-askpass pinentry-doc python3-doc python3-tk python3-venv python-dbus-doc python3-dbus-dbg python3.6-venv python3.6-doc binutils
  binfmt-support readline-doc zenity | kdialog systemd-container policykit-1
The following NEW packages will be installed:
  apparmor ca-certificates dbus dirmngr dmsetup file gir1.2-glib-2.0 gnupg gnupg-l10n gnupg-utils gpg gpg-agent gpg-wks-client gpg-wks-server gpgconf gpgsm krb5-locales
  libapparmor1 libargon2-0 libasn1-8-heimdal libassuan0 libbsd0 libcap2 libcryptsetup12 libdbus-1-3 libdevmapper1.02.1 libedit2 libexpat1 libfuse2 libgirepository-1.0-1
  libglib2.0-0 libglib2.0-data libgssapi-krb5-2 libgssapi3-heimdal libhcrypto4-heimdal libheimbase1-heimdal libheimntlm0-heimdal libhx509-5-heimdal libicu60 libidn11
  libip4tc0 libjson-c3 libk5crypto3 libkeyutils1 libkmod2 libkrb5-26-heimdal libkrb5-3 libkrb5support0 libksba8 libldap-2.4-2 libldap-common liblzo2-2 libmagic-mgc
  libmagic1 libmpdec2 libnpth0 libnss-systemd libpam-systemd libpython3-stdlib libpython3.6-minimal libpython3.6-stdlib libreadline7 libroken18-heimdal libsasl2-2
  libsasl2-modules libsasl2-modules-db libsqlite3-0 libssl1.0.0 libssl1.1 libwind0-heimdal libx11-6 libx11-data libxau6 libxcb1 libxdmcp6 libxext6 libxml2 libxmuu1
  mime-support multiarch-support networkd-dispatcher openssh-client openssl pinentry-curses python3 python3-dbus python3-gi python3-minimal python3.6 python3.6-minimal
  readline-common shared-mime-info snapd squashfs-tools systemd systemd-sysv udev xauth xdg-user-dirs xz-utils
0 upgraded, 100 newly installed, 0 to remove and 1 not upgraded.
Need to get 42.3 MB of archives.
After this operation, 187 MB of additional disk space will be used.
Do you want to continue? [Y/n]
```